### PR TITLE
Impl token whitelisting

### LINKED ISSUE #478: Make 'Claim Withdrawal' transaction feeless

### PR DESCRIPTION
**Pull request by zktony**
Impl token whitelisting. Will close #478 https://github.com/Polkadex-Substrate/orderbook/issues/375

### CODE REVIEW

**Review by Gauthamastro — DISMISSED** [review]

**Comment by mt-gareth** [comment]
Should an error be thrown if the a token that is already whitelisted is added again? Or if you try to remove a token that is not on the list?

**Comment by zktony** [comment]
> Should an error be thrown if the a token that is already whitelisted is added again? Or if you try to remove a token that is not on the list?

I intentionally did it to save computation. Anyway, storage is of type BTreeSet so it can't have duplicate items.

**Comment by mt-gareth** [comment]
> 

Great, that’s what I thought, just didn’t know if a specific error should be thrown in either case, but I get that it takes computation.

**Comment by zktony** [comment]
Review it after merging #499

**Review by serhii-temchenko — DISMISSED** [review]

File: pallets/asset-handler/src/lib.rs [hunk 1]
```diff
@@ -70,6 +70,13 @@ pub mod pallet {
 		}
 	}
 
+	pub struct WhitelistedTokenLimit;
```
Comment (35359595) [inline pallets/asset-handler/src/lib.rs:73]: should this be part of pallet's `Config` and set up from runtime?

File: pallets/asset-handler/src/lib.rs [hunk 2]
```diff
@@ -142,6 +155,10 @@ pub mod pallet {
 		BlocksDelayUpdated(T::BlockNumber),
 		/// FungibleTransferFailed
 		FungibleTransferFailed,
+		/// TokenWhitelisted
+		TokenWhitelisted(H160),
```
Comment (35359595) [inline pallets/asset-handler/src/lib.rs:159]: should we make those two similar? like `WhitelistedTokenAdded` and `WhitelistedTokenRemoved`?

File: pallets/asset-handler/src/lib.rs [hunk 3]
```diff
@@ -165,6 +182,12 @@ pub mod pallet {
 		ConversionIssue,
 		/// BridgeDeactivated
 		BridgeDeactivated,
+		/// WhitelistedTokenLimitReached
+		WhitelistedTokenLimitReached,
+		/// TokenNotWhitelisted
+		TokenNotWhitelisted,
+		/// WhitelistedTokenRemoved
```
Comment (35359595) [inline pallets/asset-handler/src/lib.rs:189]: we should start writing more descriptive docs, aspecialy here - they'll be translated into polkadot ui errors. for example: `/// This token was whitelisted but got removed and is not valid anymore`

File: pallets/asset-handler/src/lib.rs [hunk 4]
```diff
@@ -266,6 +289,30 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// Whitelist Token
+		#[pallet::weight((195_000_000).saturating_add(T::DbWeight::get().writes(1 as Weight)))]
```
Comment (35359595) [inline pallets/asset-handler/src/lib.rs:293]: this extrinsic does one read and two writes in works case, which should be reflected

Comment (35359595) [reply pallets/asset-handler/src/lib.rs:293]: ah no, it does one read and one write. please re-write using `.mutate()`

File: pallets/asset-handler/src/lib.rs [hunk 5]
```diff
@@ -266,6 +289,30 @@ pub mod pallet {
...
+			Self::deposit_event(Event::<T>::TokenWhitelisted(token_add));
+			Ok(())
+		}
+
+		/// Remove Whitelisted Token
+		#[pallet::weight((195_000_000).saturating_add(T::DbWeight::get().writes(1 as Weight)))]
+		pub fn remove_whitelisted_token(origin: OriginFor<T>, token_add: H160) -> DispatchResult {
+			T::AssetCreateUpdateOrigin::ensure_origin(origin)?;
+			let mut whitelisted_tokens = <WhitelistedToken<T>>::get();
+			whitelisted_tokens.remove(&token_add);
```
Comment (35359595) [inline pallets/asset-handler/src/lib.rs:310]: also re-implement this using `mutate()`

File: pallets/chainbridge/src/lib.rs [hunk 6]
```diff
@@ -27,7 +27,7 @@ pub mod pallet {
 	use sp_std::prelude::*;
 
 	const DEFAULT_RELAYER_THRESHOLD: u32 = 1;
-	const MODULE_ID: PalletId = PalletId(*b"phala/bg");
+	const MODULE_ID: PalletId = PalletId(*b"polka/bg");
```
Comment (35359595) [inline pallets/chainbridge/src/lib.rs:30]: I propose more polkadex name like `pdex/bg`

File: pallets/chainbridge/src/mock.rs [hunk 7]
```diff
@@ -108,7 +108,7 @@ pub const ENDOWED_BALANCE: u64 = 100_000_000;
 pub const TEST_THRESHOLD: u32 = 2;
 
 pub fn new_test_ext() -> sp_io::TestExternalities {
-	let bridge_id = PalletId(*b"polka/bg").into_account_truncating();
+	let bridge_id: u64 = PalletId(*b"polka/bg").into_account_truncating();
```
Comment (35359595) [inline pallets/chainbridge/src/mock.rs:111]: same proposal -> `pdex/bg`

File: pallets/ocex/src/lib.rs [hunk 8]
```diff
@@ -159,19 +163,29 @@ pub mod pallet {
 
 	#[pallet::error]
 	pub enum Error<T> {
+		///RegisterationShouldBeSignedByMainAccount
```
Comment (35359595) [inline pallets/ocex/src/lib.rs:166]: would love to see more descriptive message here - something similar to old errors below with proper description would be nice

File: pallets/ocex/src/lib.rs [hunk 9]
```diff
@@ -791,7 +810,8 @@ impl<T: Config> Pallet<T> {
 				)?;
 			},
 			AssetId::asset(id) => {
-				T::OtherAssets::teleport(id, payer, payee, amount.unique_saturated_into())?;
+				//panic!();
```
Comment (35359595) [inline pallets/ocex/src/lib.rs:813]: remove this commented code

File: pallets/ocex/src/types.rs [hunk 10]
```diff
@@ -30,3 +30,5 @@ impl TradingPairStatus {
 		Self { is_active: true }
 	}
 }
+
```
Comment (35359595) [inline pallets/ocex/src/types.rs:33]: remove these two extra empty lines

File: pallets/polkadex-ido/src/lib.rs [hunk 11]
```diff
@@ -101,11 +101,11 @@ pub mod weights;
 use pallet_polkadex_ido_primitives::{FundingRoundWithPrimitives, StringAssetId, VoteStat};
 pub use weights::WeightInfo;
 
-#[cfg(test)]
-mod mock;
-
-#[cfg(test)]
-mod test;
+// #[cfg(test)]
+// mod mock;
```
Comment (35359595) [inline pallets/polkadex-ido/src/lib.rs:105]: why commenting this?

File: pallets/polkadex-ido/src/lib.rs [hunk 12]
```diff
@@ -101,11 +101,11 @@ pub mod weights;
...
+//
+// #[cfg(test)]
+// mod test;
```
Comment (35359595) [inline pallets/polkadex-ido/src/lib.rs:108]: why commenting this?

File: pallets/test-token-provider/src/lib.rs [hunk 13]
```diff
@@ -3,8 +3,8 @@
 #[cfg(test)]
 mod mock;
 
-#[cfg(test)]
-mod test;
+// #[cfg(test)]
+// mod test;
```
Comment (35359595) [inline pallets/test-token-provider/src/lib.rs:7]: why commenting this?

File: pallets/test-token-provider/src/lib.rs [hunk 14]
```diff
@@ -3,8 +3,8 @@
...
 
 pub use pallet::*;
 // use sp_core::H160;
```
Comment (35359595) [inline pallets/test-token-provider/src/lib.rs:10]: remove commented code

**Review by 35359595 — CHANGES_REQUESTED** [review]

File: pallets/polkadex-ido/src/lib.rs (hunk 11 above)
Comment (zktony) [reply pallets/polkadex-ido/src/lib.rs:105]: We are going to revamp IDO Pallet completely. So there is no need to fix tests of the old IDO Pallet. As we are going to remove it.

File: pallets/polkadex-ido/src/lib.rs (hunk 12 above)
Comment (zktony) [reply pallets/polkadex-ido/src/lib.rs:108]: We are going to revamp IDO Pallet completely. So there is no need to fix tests of the old IDO Pallet. As we are going to remove it.

File: pallets/test-token-provider/src/lib.rs (hunk 13 above)
Comment (zktony) [reply pallets/test-token-provider/src/lib.rs:7]: This pallet is just meant for testing. We are going to remove it eventually.

File: pallets/test-token-provider/src/lib.rs (hunk 14 above)
Comment (zktony) [reply pallets/test-token-provider/src/lib.rs:10]: This pallet is just meant for testing. We are going to remove it eventually.

**Comment by serhii-temchenko** [comment]
@zktony is it ready to be reviewed? I see that it has `InProgress` label and at the same time @35359595 already reviewed it so I'm a bit confused. :)

File: pallets/asset-handler/src/lib.rs (hunk 1 above)
Comment (zktony) [reply pallets/asset-handler/src/lib.rs:73]: I tried to do that but then I was not able to use it inside storage. So I think we cant us inside config.

**Comment by zktony** [comment]
Resolved this in different PR #523

**Pull request closed without merging**